Geyser Runtime Reload

cavemanloverboy opened the pull request:
#### Problem
During the typical geyser plugin dev cycle, the geyser plugin often needs to be rebuilt and redeployed. In the current `solana-validator`, this requires bringing the validator client down, and pointing the `GeyserPluginManager` to a new plugin that it loads at runtime anyway.

This is a problem for multiple reasons. The first is the obvious one, which is that a dev must often wait for their validator to catch up to the tip, which can slow down development. The second is that this has a minor impact on Solana uptime. If we can make Geyser plugins reloadable, this would alleviate both of these problems.

#### Summary of Changes
This set of changes adds a set of threads which periodically (every 10s) check for a `refresh_yourplugin` file in the same directory as your plugin. Upon detecting such a file, the thread locks the now-shared state of the `GeyserPluginManager`, attempts to load and initialize the plugin. If successful, the old `Box<dyn GeyserPlugin>` is shut down and then overwritten.

These changes allow a dev to replace a currently running `GeyserPlugin` without having to bring down the validator. To do so, a dev need only
```
touch refresh_yourplugin
```
in the same directory as the plugin and wait 10s. Within 10s, this file will be replaced with a result_refresh_yourplugin which notifies the dev of the result of the change. This can be one of the following:
```
"success"
"created plugin but on_load failed"
"failed to create_plugin"
"failed to find plugin"
```
This could be done instead by listening to some port but this brings about particular security concerns if the dev doesn't have a particular port locked down. In addition, it is likely easier to reload the wrong geyser plugin (if multiple plugins are used) if pinging the same port for every plugin.

The code works, but is not optimal by any means.

cavemanloverboy commented:
shoutout: inspiration from this came from 

https://enzo.readthedocs.io/en/latest/user_guide/ControllingDataOutput.html

```
touch outputNow
```

cavemanloverboy commented:
The only thing I am iffy about is the ordering of the shutdown/startup calls. 

If you shutdown the old plugin first before trying to start up the second, then you risk the startup command of the new plugin failing and then you don't have any geyser plugin. However, if there's anything used by the plugin (e.g. binds to a particular port), then the new plugin start up might fail if you try to start the new one before shutting the old one down.

lijunwangs commented on geyser-plugin-manager/src/accounts_update_notifier.rs:
```diff
@@ -74,6 +74,8 @@ impl AccountsUpdateNotifierInterface for AccountsUpdateNotifierImpl {
         }
 
         for plugin in plugin_manager.plugins.iter_mut() {
+            let mut plugin = plugin.write().unwrap();
```
Do not take write lock, it will hurt concurrency. Take the read lock

cavemanloverboy replied:
All of the relevant methods in `GeyserPlugin` take a `&mut self`, for example, here is the `update_account` signature:

```rust
    #[allow(unused_variables)]
    fn update_account(
        &mut self,
        account: ReplicaAccountInfoVersions,
        slot: u64,
        is_startup: bool,
    ) -> Result<()> {
        Ok(())
    }
   ```
   
   https://github.com/solana-labs/solana/blob/4639919c9b087b68bc24a736fbed26105447b011/geyser-plugin-interface/src/geyser_plugin_interface.rs#L265-L277

lijunwangs replied:
Right -- we need to think carefully. mut meaning is mutable -- it does not mean it is mutually exclusive -- concurrent operations can happen to a mut object and it is happening.

cavemanloverboy replied:
The code isn't going to compile with a read lock. There can exist only one mutable reference for a type at any given time.

cavemanloverboy replied:
But I think what you might be going for or thinking about is that we've already (just) taken a lock on the plugin manager so we shouldn't also need a lock on the plugin.

lijunwangs replied:
Good point. I need to review how lock is being done overall in geyser. We might be hurting concurrency unnecessarily.

cavemanloverboy replied:
I don't think there's a way to make concurrent calls into the workhorse functions of a `GeyserPlugin` without breaking changes. You would have to require that the plugin take in `&self` on all methods and require plugin writers to use interior mutability to update plugin state. Maybe a `GeyserPluginV2` can come online down the road. With the current trait interface, locks are needed.

Taking a lock twice is mildly expensive, but not prohibitively so. 1 million double locks is 20-40ms on whatever hardware the playground uses.

https://play.rust-lang.org/?version=stable&mode=release&edition=2021&gist=4fde2ff5764268dae083e2a53ab364e6

cavemanloverboy replied:
I also don't know if there are any assumptions baked into the geyser plugin, such as account state changes being received in consensus order. Adding a concurrent-friendly interface to that may break those assumptions.

lijunwangs replied:
Sound right. Keep this code for now. I will address the overall locking separately. Yes, I might have to break the interface, but it is necessary. We have found validator cannot keep up with the network with plugin loaded.

nickgarfield replied:
We've been seeing our nodes (RPC) fall behind as well. They have plenty of cpu/mem/bandwidth capacity available, so it's been very confusing to debug. This sequential write lock explanation makes sense. 

@lijunwangs would love to help in any way we can to resolve this.

nickgarfield replied:
@lijunwangs Would love if you could offer any feedback on this one. I was able to get a `solana-test-validator` running locally with a plugin and didn't notice any issues. I'm going to proceed by cherry-picking back onto v1.14 and deploying a canary to devnet and testnet. 

https://github.com/solana-labs/solana/pull/30380

lijunwangs replied:
I think we can table this lock issue now for this PR. I will address it separately. My major concern is how the system detect the need to reload. I would like to implement via the admin RPC. So the flow likes the following

User ---> Admin RPC --> Send (load/reload) request --> Plugin Manager --> Load/Reload a plugin.

This way we do not have to busily scan for reload which is rare and once it is requested we can quickly handle it. In addition the same mechanism can be used to load a new plugin.

cavemanloverboy replied:
Yea that sounds great. Are there any admin RPC methods I could use a reference? Can I get a permalink?

lijunwangs replied:
cc. @t-nelson @CriesofCarrots

cavemanloverboy commented:
It might be good to have a single thread that iterates over multiple plugins instead of a single thread for each plugin

lijunwangs commented on geyser-plugin-manager/src/geyser_plugin_manager.rs:
```diff
@@ -29,21 +38,115 @@ impl GeyserPluginManager {
         let constructor: Symbol<PluginConstructor> = lib.get(b"_create_plugin")?;
         let plugin_raw = constructor();
-        let mut plugin = Box::from_raw(plugin_raw);
+        let mut plugin: Box<dyn GeyserPlugin> = Box::from_raw(plugin_raw);
```
Is this change needed? Avoid unnecessary changes if not

cavemanloverboy replied:
Not needed. I'll remove it

lijunwangs commented on geyser-plugin-manager/src/geyser_plugin_manager.rs:
```diff
@@ -29,21 +38,115 @@ impl GeyserPluginManager {
+
+        let owned_path = std::path::PathBuf::from(libpath);
+        let owned_config = config_file.to_owned();
+        let handle = std::thread::spawn(move || {
```
Spawn a thread for each plugin is an overkill. I think we do not want to do this busily. This can be better handled by an administrative RPC call, in the RPC the user can direct reload or load a new plugin. So that it can solve both use cases.

cavemanloverboy replied:
administrative RPC sounds like a much better solution.

t-nelson commented on geyser-plugin-manager/src/geyser_plugin_manager.rs:
```diff
@@ -29,21 +38,115 @@ impl GeyserPluginManager {
+                refresh_path
+                    .file_name()
+                    .map(std::ffi::OsStr::to_str)
+                    .expect("valid path with valid filename was given if library loaded")
```
every `expect()` in the function need to go.  user input validation needs to be fallible without taking down the process

t-nelson commented on geyser-plugin-manager/src/geyser_plugin_manager.rs:
```diff
@@ -29,21 +38,115 @@ impl GeyserPluginManager {
+                    .expect("valid path with valid filename was given if library loaded")
+            ));
+
+            #[allow(unused_must_use)] // result io
```
no!

t-nelson commented on geyser-plugin-manager/src/geyser_plugin_manager.rs:
```diff
@@ -29,21 +38,115 @@ impl GeyserPluginManager {
+            #[allow(unused_must_use)] // result io
+            loop {
+                // Every 10 seconds..
+                std::thread::park_timeout(Duration::from_secs(10));
+
+                info!("checking for new plugin at {}", refresh_path.display());
+
+                // If refresh path exists, handle refresh request
+                if refresh_path.exists() {
+                    info!("detected refresh_path {}", refresh_path.display());
```
use a channel instead

cavemanloverboy replied:
instead of a park_timeout? That's a full-on block unless it's async

t-nelson commented on geyser-plugin-manager/src/geyser_plugin_manager.rs:
```diff
@@ -29,21 +38,115 @@ impl GeyserPluginManager {
+                    info!("detected refresh_path {}", refresh_path.display());
+
+                    // Initialize file to write result
+                    let mut done_file = std::fs::File::create(&done_path)
```
what is this for?

t-nelson reviewed (commented):
this seems to only supports reloading plugins declared on process start?  it'd be far more valuable to support arbitrary unload/load/reload and i don't see how this is on the path to that

cavemanloverboy commented:
> this seems to only supports reloading plugins declared on process start?  it'd be far more valuable to support arbitrary unload/load/reload and i don't see how this is on the path to that

I agree it is more powerful. Point me to an administrative RPC call and I'll submit a new PR tomorrow

t-nelson commented:
> > this seems to only supports reloading plugins declared on process start?  it'd be far more valuable to support arbitrary unload/load/reload and i don't see how this is on the path to that
> 
> I agree it is more powerful. Point me to an administrative RPC call and I'll submit a new PR tomorrow. I wasn't sure if you all would be receptive so I just hacked this together.

admin rpc service is defined here, https://github.com/solana-labs/solana/blob/master/validator/src/admin_rpc_service.rs :saluting_face:

cavemanloverboy commented:
@t-nelson I wrote a reload geyser admin rpc method with a test. If its up to your standards, I will write the other methods.

If you want tests in the geyser plugin interface and manager crates (both have zero tests at the moment), you could copy this dummy plugin/library.

t-nelson commented on validator/src/admin_rpc_service.rs:
```diff
@@ -48,6 +55,7 @@ pub struct AdminRpcRequestMetadata {
     pub tower_storage: Arc<dyn TowerStorage>,
     pub staked_nodes_overrides: Arc<RwLock<HashMap<Pubkey, u64>>>,
     pub post_init: Arc<RwLock<Option<AdminRpcRequestMetadataPostInit>>>,
+    pub plugin_manager: Arc<RwLock<GeyserPluginManager>>,
```
admin service doesn't need a reference to the plugin manager.  instead, plugin manager and admin services should share (opposing ends of) a pair of [channels](https://doc.rust-lang.org/std/sync/mpsc/fn.channel.html).  plugin manager now gets a request receiver and a reply sender on init, spools up a thread and waits for requests to service. admin service gets the request sender and reply receiver, then forwards requests and replies via the admin socket.

this will allow us to encapsulate all of the plugin logic in the plugin manager where it belongs, rather than scatter it all over the place. we also avoid exposing plugin manager internals that don't need to be (`pub` fields are evil)

cavemanloverboy replied:
I urge you to consider the benefits of the current setup over encapsulation. 

1) Save an entire thread spawn + having to periodically check for updates. As is, this doesn't affect the performance of the plugin_manager when the admin rpc service is unused. The admin rpc service is currently given 3 worker threads*** so this would be putting those to better use.
2) better reload latency vs the additional channel send/recv pings.
3) the `pub libs` aren't actually used for anything after startup and should actually probably just be dropped after calling `lib.get(b"_create_plugin")?` when loading a plugin. The only thing a lib can do is call `lib.get(b"_create_plugin")?` which should put the whole vtable in memory. As I'm writing this, i'm realizing that if we choose to keep them for whatever reason we may want to consider rebuilding the old plugin by calling `current_lib.get(b"_create_plugin")?` and then calling `on_load` versus just calling `on_load`.
4) It is possible to get a bit more encapsulation without additional thread spawns by writing the reload logic as a method of the plugin manager which is called in the admin rpc service.

*** 1) What

lijunwangs replied:
Agreed with @t-nelson : no need to build direct dependency of admin_rpc_service on plugin manager. It only needs a send handle to the channel.

cavemanloverboy replied:
You are asking the admin RPC service to manage plugins; instead of adding a dependency on the plugin manager***, you would rather have a runtime performance hit with thread spawn + channels? 

***the admin rpc service is defined in `solana-validator` which is already downstream from `solana-core` which which is already downstream from `solana-geyser-plugin-manager`. This isn't really adding a new dependency. We will always have
```
solana-validator v1.16.0
├── solana-core v1.16.0
│   ├── solana-geyser-plugin-manager v1.16.0
```
unless you plan to rip the manager out of `solana-core`.

cavemanloverboy replied:
If you **really really** want another thread to do this, I will concede upon confirmation. But I think it's a bad idea.

lijunwangs replied:
> If you **really really** want another thread to do this, I will concede upon confirmation. But I think it's a bad idea.

Yes. That is a widely used pattern in solana validator to break reduce dependencies across modules. The administrator RPC does not need to know who is actually handling that event. Suppose in one day I have wide user use case and may need to change to another handler -- we do not need to change the RPC code at all.  As long as the thread waiting for the event with good timeout like recv_timeout I do not think any performance penalties -- the OS should be able do good scheduling to use the CPU for such light activity. In addition, we do not want rpc handler to handle this synchronously any way as the operation can takes sometime.

lijunwangs replied:
I mentioned this approach in the comment https://github.com/solana-labs/solana/pull/30352#issuecomment-1435062751

cavemanloverboy replied:
If I define some enum (see below), in which crate do I put it to minimize dependencies? If I put it in the manager plugin crate, then the validator is still dependent on the manager plugin crate. Do I put it in some existing common or utils crate? If so, can you point me to it?
```rust
pub enum PluginManagerRequest {
    ReloadPlugin {
        name: String,
        libpath: String,
        config_file: String,
    },
    ListPlugins,
}
```

CriesofCarrots replied:
Just chiming in to agree with Trent and Lijun. I do still hope to disentangle the admin-rpc-service from the solana-validator crate one day, and while there are a couple wrinkles to doing so already, I’d rather not add another one.

lijunwangs replied:
Yes, we need to define the event type regardless and I would just define it pub enum PluginRequest and define in geyser-plugin-manager crate and put it in geyser_plugin_service.rs. This is still better than expose the fat class GeyserPluginManager out. You can see the existing code out side of geyser-plugin-manager do not need to know its implementation.

CriesofCarrots replied:
> and define in geyser-plugin-manager crate

We might want to create a new `solana-admin-rpc-api` crate to hold types like this. Wdyt, @lijunwangs ?

cavemanloverboy replied:
If introducing a new crate, I think it would be best to position it to hold other common types. Maybe call it solana-common

t-nelson reviewed (dismissed):
this is much closer!  i left a recommendation to improve encapsulation

lijunwangs commented on validator/src/admin_rpc_service.rs:
```diff
@@ -146,6 +152,14 @@ pub trait AdminRpc {
+    fn reload_plugin(
+        &self,
+        meta: Self::Metadata,
+        libpath: String,
```
We should only requires the config_file -- we have already documented the config file must have the "libpath" in its json content.

cavemanloverboy replied:
the user might have moved the geyser plugin.

cavemanloverboy replied:
marking as resolved as rpc methods only require config file

cavemanloverboy commented on geyser-plugin-manager/src/geyser_plugin_manager.rs:
```diff
@@ -72,4 +78,270 @@ impl GeyserPluginManager {
+
+        // Try unload, on_load
+        // Attempt to load new plugin
+        match new_plugin.on_load(&config_file) {
```
Yea I've rewritten this piece. I will push up the latest version by EoD

cavemanloverboy replied:
marking as resolved as there is no attempt to revert now

cavemanloverboy commented:
> Yes, we need to define the event type regardless and I would just define it pub enum PluginRequest and define in geyser-plugin-manager crate and put it in geyser_plugin_service.rs. This is still better than expose the fat class GeyserPluginManager out. You can see the existing code out side of geyser-plugin-manager do not need to know its implementation.

@lijunwangs I can't seem to respond to one of your comments so I'm responding here. The admin rpc service needs to use the type in order to forward the request to the plugin manager thread.

lijunwangs commented:
> > Yes, we need to define the event type regardless and I would just define it pub enum PluginRequest and define in geyser-plugin-manager crate and put it in geyser_plugin_service.rs. This is still better than expose the fat class GeyserPluginManager out. You can see the existing code out side of geyser-plugin-manager do not need to know its implementation.
> 
> @lijunwangs I can't seem to respond to one of your comments so I'm responding here. The admin rpc service needs to use the type in order to forward the request to the plugin manager thread.

I understand what you meant and agreed we need to define the request type. The above is what I thought where we should define it.

cavemanloverboy commented:
> > > Yes, we need to define the event type regardless and I would just define it pub enum PluginRequest and define in geyser-plugin-manager crate and put it in geyser_plugin_service.rs. This is still better than expose the fat class GeyserPluginManager out. You can see the existing code out side of geyser-plugin-manager do not need to know its implementation.
> 
> > 
> 
> > @lijunwangs I can't seem to respond to one of your comments so I'm responding here. The admin rpc service needs to use the type in order to forward the request to the plugin manager thread.
> 
> 
> 
> I understand what you meant and agreed we need to define the request type. The above is what I thought where we should define it.

Wait so you want it in the geyser plugin manager crate? Doesn't that require the dependency on the manager crate which is the thing we didn't want?

cavemanloverboy commented on geyser-plugin-manager/src/geyser_plugin_manager.rs:
```diff
@@ -72,4 +78,270 @@ impl GeyserPluginManager {
+        // Check if any plugin names match this one
+        let Some(idx) = plugin_manager.plugins.iter().position(|plugin| plugin.name().eq(name)) else {
+            // If we don't find one, drop write lock ASAP and return an error
+            drop(plugin_manager);
```
This drops it one line earlier, explicitly before the return, so that the other threads can continue sending notifies a few cycles earlier.

cavemanloverboy commented on geyser-plugin-manager/src/geyser_plugin_manager.rs:
```diff
@@ -72,4 +78,270 @@ impl GeyserPluginManager {
+            #[cfg(not(test))]
+            unsafe {
+                // Attempt to load Library
+                let lib = Library::new(libpath).map_err(|e| {
```
The existing code doesn't return concrete error types. I want to be able to return an informative and explicit error saying either that the library was not found or that the create plugin symbol was not found. If the libloading   crate gives good errors I suppose I can just propagate. Im also not a fan of dynamic dispatch generally, especially when not needed.

lijunwangs replied:
That is why I said -- "refactor", instead of duplicating logic, refactor and share code

lijunwangs commented:
> 

We need to define the request type regardless -- we could put it in a separate mod which we do not have and overkill to introduce. Let's just put in the place I mentioned above. It is still different from asking the RPC code to directly work with GeyserPluginManager.

cavemanloverboy commented:
> > 
> 
> 
> 
> We need to define the request type regardless -- we could put it in a separate mod which we do not have and overkill to introduce. Let's just put in the place I mentioned above. It is still different from asking the RPC code to directly work with GeyserPluginManager. 

Sounds good

cavemanloverboy commented:
I've implemented the geyser plugin runtime reload using channels and a thread which is parked until it's time to change plugins. I removed the `libs` and `libpaths` from `GeyserPluginManager`, as those are not used at all, ever, anywhere, after loading plugins.

I've also added a few commands to the solana-validator cli to help execute these commands. 

Having spun up a `solana-test-validator` with [this test plugin](https://github.com/mwrites/solana-geyser-plugin-scaffold), I am able to execute

```
$ solana-validator --ledger=test-ledger list
Currently the following plugins are loaded:
  1) GeyserPluginHook

$ solana-validator --ledger=test-ledger plugin unload GeyserPluginHook
Successfully unloaded plugin: GeyserPluginHook

$ solana-validator --ledger=test-ledger list
There are currently no plugins loaded

$ solana-validator --ledger=test-ledger plugin load ./target/release/libsolana_geyser_plugin_scaffold.dylib config/geyser-plugin-config-mac.json
Successfully loaded plugin: GeyserPluginHook

$ solana-validator --ledger=test-ledger list
Currently the following plugins are loaded:
  1) GeyserPluginHook

$ solana-validator --ledger=test-ledger plugin reload --name=GeyserPluginHook --lib=./target/release/libsolana_geyser_plugin_scaffold.dylib --config=config/geyser-plugin-config-mac.json
Successfully reloaded plugin: GeyserPluginHook

$ solana-validator --ledger=test-ledger list
Currently the following plugins are loaded:
  1) GeyserPluginHook
```

I also checked that I get errors such as 
```
$ solana-validator --ledger=test-ledger plugin load ./target/release/libsolana_geyser_plugin_scaffold.dylib config/geyser-plugin-config-mac.json
failed to load plugin config/geyser-plugin-config-mac.json: JsonRpcError(Error { code: InternalError, message: "There already exists a plugin GeyserPluginHook loaded. Did not load requested plugin", data: None })
```
when loading a plugin whose name already matches one loaded, and similar ones for unloading or reloading a plugin that doesn't exist. 

I repeated the experiment 10 times and get the behavior I expect.

In case you are wondering why I think reload is needed and not redundant with unload + load, the reload essentially atomically takes a write lock and updates the plugin, so the plugin will not miss a single transaction, block, slot, or account update.

CriesofCarrots commented on geyser-plugin-manager/Cargo.toml:
```diff
@@ -13,6 +13,8 @@ documentation = "https://docs.rs/solana-geyser-plugin-manager"
 bs58 = "0.4.0"
 crossbeam-channel = "0.5"
 json5 = "0.4.1"
+jsonrpc-core = "18.0.0"
+jsonrpc-server-utils = "18.0.0"
```
It would be best not to add jsonrpc dependency here. You can add a custom error variant for the Result.
Is there a reason we can't use tokio directly for the channel?

cavemanloverboy replied:
Tokio is not a direct dependency. I could add it.

cavemanloverboy replied:
I had actually planned to add another error variant, but that was before we said it was okay to add a dependency for the Request enum. I could add it now.

CriesofCarrots commented on validator/src/admin_rpc_service.rs:
```diff
@@ -241,6 +271,177 @@ impl AdminRpc for AdminRpcImpl {
+    ) -> BoxFuture<Result<()>> {
+        Box::pin(async move {
+            // Construct channel for plugin to respond to this particular rpc request instance
+            let (tx, rx) = oneshot_channel();
```
```suggestion
            let (plugin_result_sender, plugin_result_receiver) = oneshot_channel();
```
`tx` usually means transaction in our code base.
Generally, please spell things out.

cavemanloverboy replied:
Makes sense. Will do

cavemanloverboy replied:
marked as resolved, as i've changed all tx, rx to sender and receiver

cavemanloverboy commented:
I moved the RPC methods in to `GeyserPluginService` since that is where the rpc handler thread and the config_file parser is defined.

lijunwangs commented on geyser-plugin-manager/src/geyser_plugin_manager.rs:
```diff
@@ -73,3 +64,47 @@ impl GeyserPluginManager {
         false
     }
 }
+
```
Why adding new code commented out? Please mark the PR in draft mode and turn it back on if it is ready for review?

cavemanloverboy replied:
Oops, forgot to clean this up. All the code that isn't commented is what I want a review on.

cavemanloverboy commented:
One final change that we may want: right now a `GeyserPluginService` is only spun up if at least one config flag is passed at initialization. If a validator wants to allow for loading plugins later, we could provide an optional flag `--geyser-service` which spins up the service with zero plugin, allowing for plugin management later.

lijunwangs commented on geyser-plugin-manager/src/geyser_plugin_service.rs:
```diff
@@ -208,4 +221,384 @@ impl GeyserPluginService {
+}
+
+#[cfg(not(test))]
+fn load_plugin_from_config(
```
Put this code near where the load_plugin was to make it easier to review -- there is not much real change.

lijunwangs commented on geyser-plugin-manager/src/geyser_plugin_service.rs:
```diff
@@ -208,4 +221,384 @@ impl GeyserPluginService {
+            })?;
+
+        // Then see if a plugin with this name already exists. If so, abort
+        if plugin_manager
```
Why are we loading first and check later? Would not you find yourself?

cavemanloverboy replied:
To get the name of the plugin, you need to call the `name` method which can only be called once you load it.

lijunwangs commented on geyser-plugin-manager/src/geyser_plugin_service.rs:
```diff
@@ -208,4 +221,384 @@ impl GeyserPluginService {
+        name: &str,
+    ) -> JsonRpcResult<()> {
+        // Check if any plugin names match this one
+        let Some(idx) = plugin_manager.plugins.iter().position(|plugin| plugin.name().eq(name)) else {
```
As mentioned do not directly modify the plugins under plugin manager -- these logic belongs to plugin manager.

lijunwangs commented on validator/src/cli.rs:
```diff
@@ -1468,6 +1468,48 @@ pub fn app<'a>(version: &'a str, default_args: &'a DefaultArgs) -> App<'a, 'a> {
         )
+        .subcommand(
+            SubCommand::with_name("plugin")
+                .about("Manage and view plugins")
```
plugins --> Geyser plugins

lijunwangs replied:
Again only here, please clarify by adding "Geyser"

cavemanloverboy commented on geyser-plugin-manager/src/geyser_plugin_service.rs:
```diff
@@ -122,70 +128,77 @@ impl GeyserPluginService {
+                            let plugin_manager = rpc_plugin_manager.read().unwrap();
+                            response_sender
+                                .send(Self::list_plugins_rpc(&plugin_manager))
+                                .expect("admin rpc service will be waiting for response");
```
Idiomatically, an `expect` message is supposed to contain a message explaining why it should never fail, not that it failed. See [this](https://doc.rust-lang.org/std/result/enum.Result.html#method.expect)

In this case, the send will succeed because rpc service will not drop the receiver and will be waiting for the response.

cavemanloverboy commented on geyser-plugin-manager/src/geyser_plugin_service.rs:
```diff
@@ -122,70 +128,77 @@ impl GeyserPluginService {
             (None, None)
         };
 
+        let rpc_handler_thread = rpc_to_plugin_manager_receiver.map(|request_receiver| {
```
good call

cavemanloverboy commented on geyser-plugin-manager/src/geyser_plugin_manager.rs:
```diff
@@ -47,10 +42,6 @@ impl GeyserPluginManager {
             plugin.on_unload();
         }
-
-        for lib in self.libs.drain(..) {
```
I will revert this and submit another pr for this after this one is finalized then.

cavemanloverboy commented on geyser-plugin-manager/src/geyser_plugin_service.rs:
```diff
@@ -122,70 +128,77 @@ impl GeyserPluginService {
+            let rpc_plugin_manager = plugin_manager.clone();
+            Arc::new(std::thread::spawn(move || loop {
+                std::thread::park();
+                info!("we've been awakened");
```
Ah yes, this was a debug print statement. Will clean up

cavemanloverboy commented on geyser-plugin-manager/src/geyser_plugin_service.rs:
```diff
@@ -122,70 +128,77 @@ impl GeyserPluginService {
+        let rpc_handler_thread = rpc_to_plugin_manager_receiver.map(|request_receiver| {
+            let rpc_plugin_manager = plugin_manager.clone();
+            Arc::new(std::thread::spawn(move || loop {
+                std::thread::park();
```
If we remove the park we should just add a `recv()`.

Idk why but I thought the channels used spinlocks; it uses park/unpark internally.

cavemanloverboy commented on geyser-plugin-manager/src/geyser_plugin_service.rs:
```diff
@@ -122,70 +128,77 @@ impl GeyserPluginService {
-            }
-        }
+        let plugin: Box<dyn GeyserPlugin> = load_plugin_from_config(geyser_plugin_config_file)?;
+        plugin_manager.plugins.push(plugin);
```
Currently both the `GeyserPluginService` and the inner `GeyserPluginManager` are managing plugins. You currently have a `load_plugin` function in the geyser plugin service with `GeyserPluginServiceError`s being served upon failure. If you want to continue to have the `GeyserPluginService` handle the requests, then we need to be able to access the inner types in the service. If you instead want to move the handling into the manager by redirecting it to the maanger and migrate/rename `GeyserPluginServiceError` to `GeyserPluginManagerError` I can do that. That is probably much cleaner.

lijunwangs replied:
My main port -- do not modify PluginManager.plugins outside of that module. The Manager essentially is managing that container and physical load/unload of plugin. The PluginService is main interface used out side of the crate. It load_plugin only read the config content to get the libpath -- all real load is handled by manager. No need to make such change. Again, please reduce unnecessary changes.

lijunwangs commented on ledger-tool/src/main.rs:
```diff
@@ -1222,13 +1222,18 @@ fn load_bank_forks(
-            );
+
+        // solana-ledger-tool doesn't run an admin_rpc service, so just pass in a dummy rx
+        let rpc_to_plugin_manager_rx = None;
```
I think Tyera made a point before, name it to rpc_to_plugin_manager_receiver -- it is easier to read. So please avoid rx, tx acronyms here.

cavemanloverboy replied:
Hmm I thought I got them all. Missed this one

cavemanloverboy commented on validator/src/admin_rpc_service.rs:
```diff
@@ -241,6 +265,177 @@ impl AdminRpc for AdminRpcImpl {
+                        config_file,
+                        response_sender,
+                    })
+                    .expect("plugin manager should never drop request receiver");
```
Sounds good. I will fix this. As a heads up, there are plenty of error messages in the codebase that are not capitalized.

cavemanloverboy commented on core/src/validator.rs:
```diff
@@ -1112,6 +1119,12 @@ impl Validator {
+    pub fn plugin_manager_handle(&self) -> Option<Arc<JoinHandle<()>>> {
+        self.geyser_plugin_service
+            .as_ref()
+            .map(|gps| gps.plugin_manager_join_handle())
```
Since i am no longer parking the thread, there is no need to propagate the handle

cavemanloverboy commented:
1) I removed the remaining `info!` message I used for debugging
2) I removed the thread handle + park/unpark + recv_timeout
3) I properly capitalized the error messages in the admin rpc service
4) I refactored the manager rpc handler thread spawn into its own function
5)  I renamed the remaining instances of `rx` to `receiver` (there are many `tx` and `rx` in pubsub client, by the way. yet another PR).

I still need to resolve merge conflicts

lijunwangs commented:
CI is not happy -- please make sure you go to programs/sbf and do `cargo build` and check in the resultant cargo.lock file.

lijunwangs commented on geyser-plugin-manager/src/geyser_plugin_manager.rs:
```diff
@@ -72,4 +52,446 @@ impl GeyserPluginManager {
+            .any(|plugin| plugin.name().eq(new_plugin.name()))
+        {
+            return Err(jsonrpc_core::Error {
+                code: ErrorCode::InternalError,
```
InternalError --> InvalidRequest

cavemanloverboy replied:
done will push

lijunwangs commented on geyser-plugin-manager/src/geyser_plugin_manager.rs:
```diff
@@ -72,4 +52,446 @@ impl GeyserPluginManager {
+        let (mut new_plugin, new_lib, new_config_file) =
+            load_plugin_from_config(geyser_plugin_config_file.as_ref()).map_err(|e| {
+                jsonrpc_core::Error {
+                    code: ErrorCode::InternalError,
```
InternalError --> InvalidRequest

cavemanloverboy replied:
done will push

lijunwangs commented on geyser-plugin-manager/src/geyser_plugin_manager.rs:
```diff
@@ -72,4 +52,446 @@ impl GeyserPluginManager {
+            // If we don't find one return an error
+            return Err(
+                jsonrpc_core::error::Error {
+                    code: ErrorCode::InternalError,
```
InternalError --> InvalidRequest

cavemanloverboy replied:
done, will push

lijunwangs commented on geyser-plugin-manager/src/geyser_plugin_manager.rs:
```diff
@@ -72,4 +52,446 @@ impl GeyserPluginManager {
+        let (mut new_plugin, new_lib, new_config_file) =
+            load_plugin_from_config(geyser_plugin_config_file.as_ref()).map_err(|e| {
+                jsonrpc_core::Error {
+                    code: ErrorCode::InternalError,
```
I think most errors returning from that function can be classified as InvalidRequest

cavemanloverboy replied:
done. will push

lijunwangs commented on geyser-plugin-manager/src/geyser_plugin_manager.rs:
```diff
@@ -72,4 +52,446 @@ impl GeyserPluginManager {
+                return Err(jsonrpc_core::error::Error {
+                    code: ErrorCode::InternalError,
+                    message: format!(
+                        "failed to start new plugin (previous plugin was dropped!): {err}"
```
Nits: "Failed to start...."

cavemanloverboy replied:
fixed this one yesterday, forgot to mark as resolved

lijunwangs commented on geyser-plugin-manager/src/geyser_plugin_manager.rs:
```diff
@@ -72,4 +52,446 @@ impl GeyserPluginManager {
+    Ok((plugin, lib, config_file))
+}
+
+pub fn parse_config_file(
```
Get rid of duplicate logic in this function and the above

cavemanloverboy replied:
this one is actually unused. will remove

lijunwangs commented on ledger-tool/src/main.rs:
```diff
@@ -1222,13 +1222,18 @@ fn load_bank_forks(
-                },
-            );
+
+        // solana-ledger-tool doesn't run an admin_rpc service, so just pass in a dummy rx
```
rx --> receiver

cavemanloverboy replied:
done

lijunwangs commented on geyser-plugin-manager/src/geyser_plugin_service.rs:
```diff
@@ -208,4 +150,62 @@ impl GeyserPluginService {
+        plugin_manager: Arc<RwLock<GeyserPluginManager>>,
+        request_receiver: Receiver<GeyserPluginManagerRequest>,
+    ) -> Arc<JoinHandle<()>> {
+        Arc::new(thread::spawn(move || loop {
```
I believe you need to pass in the exit Atomic variable so that when validator exit it can quit promptly.

cavemanloverboy replied:
can you point me to which exit i need to clone? seems like there are a few disconnected Arc<AtomicBool> floating around

lijunwangs commented on geyser-plugin-manager/src/geyser_plugin_service.rs:
```diff
@@ -208,4 +150,62 @@ impl GeyserPluginService {
+        request_receiver: Receiver<GeyserPluginManagerRequest>,
+    ) -> Arc<JoinHandle<()>> {
+        Arc::new(thread::spawn(move || loop {
+            if let Ok(request) = request_receiver.recv() {
```
You should use recv_timeout with good time our value like a few seconds to check for the exit

lijunwangs commented on validator/src/cli.rs:
```diff
@@ -1468,6 +1468,48 @@ pub fn app<'a>(version: &'a str, default_args: &'a DefaultArgs) -> App<'a, 'a> {
+                )
+                .subcommand(
+                    SubCommand::with_name("load")
+                        .about("Load a new plugin. Specify the plugin name, and the new config/library paths. Fails if overwriting (use reload)")
```
config path

cavemanloverboy replied:
done

lijunwangs commented on validator/src/cli.rs:
```diff
@@ -1468,6 +1468,48 @@ pub fn app<'a>(version: &'a str, default_args: &'a DefaultArgs) -> App<'a, 'a> {
+                )
+                .subcommand(
+                    SubCommand::with_name("reload")
+                        .about("Reload a particular plugin. Specify the plugin name, and the new config/library paths.")
```
config path, remove the library path

cavemanloverboy replied:
done

lijunwangs commented on validator/src/main.rs:
```diff
@@ -536,6 +537,79 @@ pub fn main() {
+                    let plugins = admin_rpc_service::runtime()
+                        .block_on(async move { admin_client.await?.list_plugins().await })
+                        .unwrap_or_else(|err| {
+                            println!("failed to list plugins: {err}");
```
--> Failed to list

cavemanloverboy replied:
done, will push

lijunwangs commented on validator/src/main.rs:
```diff
@@ -536,6 +537,79 @@ pub fn main() {
+                                admin_client.await?.unload_plugin(name.clone()).await
+                            })
+                            .unwrap_or_else(|err| {
+                                println!("failed to unload plugin {name}: {err:?}");
```
--> Failed to

cavemanloverboy replied:
done, will push

lijunwangs commented on validator/src/main.rs:
```diff
@@ -536,6 +537,79 @@ pub fn main() {
+                                admin_client.await?.load_plugin(config.clone()).await
+                            })
+                            .unwrap_or_else(|err| {
+                                println!("failed to load plugin {config}: {err:?}");
```
--> Failed to load...

cavemanloverboy replied:
done, will push

lijunwangs commented on validator/src/main.rs:
```diff
@@ -536,6 +537,79 @@ pub fn main() {
+                                        .await
+                                })
+                                .unwrap_or_else(|err| {
+                                    println!("failed to reload plugin {name}: {err:?}");
```
--> Failed to ...

cavemanloverboy replied:
done, will push

lijunwangs commented on validator/src/main.rs:
```diff
@@ -1513,6 +1588,13 @@ pub fn main() {
 
     let start_progress = Arc::new(RwLock::new(ValidatorStartProgress::default()));
     let admin_service_post_init = Arc::new(RwLock::new(None));
+    let (rpc_to_plugin_manager_tx, rpc_to_plugin_manager_receiver) = if starting_with_geyser_plugins
```
rpc_to_plugin_manager_tx --> rpc_to_plugin_manager_sender

cavemanloverboy replied:
done, pushed.

cavemanloverboy replied:
done, pushed

lijunwangs reviewed (commented):
Thanks for addressing the feedback. Some more nits and issue with handling validator exit

cavemanloverboy commented on geyser-plugin-manager/src/geyser_plugin_manager.rs:
```diff
@@ -20,37 +24,13 @@ impl GeyserPluginManager {
         for mut plugin in self.plugins.drain(..) {
             info!("Unloading plugin for {:?}", plugin.name());
             plugin.on_unload();
         }
```
Forgot to revert this change. I will revert it and then later submit a PR to remove the entire libs field (libs is never used).

cavemanloverboy replied:
reverted, will push

cavemanloverboy commented on geyser-plugin-manager/src/geyser_plugin_manager.rs:
```diff
@@ -72,4 +52,446 @@ impl GeyserPluginManager {
+#[cfg(not(test))]
+pub(crate) fn load_plugin_from_config(
+    geyser_plugin_config_file: &Path,
+) -> Result<(Box<dyn GeyserPlugin>, Library, &str), GeyserPluginManagerError> {
```
done, will push up commit

cavemanloverboy commented on geyser-plugin-manager/src/geyser_plugin_manager.rs:
```diff
@@ -72,4 +52,446 @@ impl GeyserPluginManager {
+            .on_load(new_config_file)
+            .map_err(|on_load_err| jsonrpc_core::Error {
+                code: ErrorCode::InternalError,
+                message: format!("on_load method of plugin failed: {on_load_err}",),
```
done, will push

cavemanloverboy commented on geyser-plugin-manager/src/geyser_plugin_manager.rs:
```diff
@@ -72,4 +52,446 @@ impl GeyserPluginManager {
+    pub(crate) fn load_plugin(
+        &mut self,
+        geyser_plugin_config_file: impl AsRef<Path>,
+    ) -> JsonRpcResult<String> {
```
done, will push

cavemanloverboy commented on geyser-plugin-manager/src/geyser_plugin_manager.rs:
```diff
@@ -72,4 +52,446 @@ impl GeyserPluginManager {
+            return Err(
+                jsonrpc_core::error::Error {
+                    code: ErrorCode::InternalError,
+                    message: String::from("plugin requested to unload is not loaded"),
```
done, will push

cavemanloverboy commented:
seems like the last CI label didn't do anything

cavemanloverboy commented:
my apologies for issues with CI

cavemanloverboy commented:
not sure what this previous CI error is about. let me know if you need anything on my end

lijunwangs commented:
> not sure what this previous CI error is about. let me know if you need anything on my end
That was CI environmental issue. I have restarted. Can you ensure all other comments are addressed? I still see a number of them not addressed yet. Thank you!

cavemanloverboy commented on geyser-plugin-manager/src/geyser_plugin_manager.rs:
```diff
@@ -72,4 +52,446 @@ impl GeyserPluginManager {
+            // If we don't find one return an error
+            return Err(
+                jsonrpc_core::error::Error {
+                    code: ErrorCode::InternalError,
```
done will push

cavemanloverboy commented on geyser-plugin-manager/src/geyser_plugin_manager.rs:
```diff
@@ -72,4 +52,446 @@ impl GeyserPluginManager {
+            return Err(
+                jsonrpc_core::error::Error {
+                    code: ErrorCode::InternalError,
+                    message: String::from("plugin requested to reload is not loaded"),
```
done will push

cavemanloverboy commented on geyser-plugin-manager/src/geyser_plugin_manager.rs:
```diff
@@ -72,4 +52,446 @@ impl GeyserPluginManager {
+        };
+
+        // Unload and drop current plugin first in case plugin requires exclusive access to resource,
+        // such as a particular port or database.
```
done, will push

cavemanloverboy commented:
fixed everything except the timeout + exit. which exit atomicbool do I clone? can you point me to the line in `validator/`?

lijunwangs commented:
> 



> I think I fixed everything except the timeout + exit. which exit atomicbool do I clone? can you point me to the line in `validator/`? every time `Exit::default()` is called in the codebase it seems to make a new atomic bool (not clone a previous arc).
> 
> or maybe i am misunderstanding. am I supposed to make a new atomic bool which the main thread listens for?
> 
> i will push everything after this

It is initialized here: https://github.com/solana-labs/solana/blob/master/core/src/validator.rs#L482
which is the validator exit variable sent into various service threads. You would need to move around the code to pass that into GeyserPluginService.

lijunwangs commented:
The CI also complains fmt, please use the following 

cargo +nightly-2022-11-02 fmt --all

to format the code.

Also, it looks you need rebase with the master again...

cavemanloverboy commented:
> The CI also complains fmt, please use the following
> 
> cargo +nightly-2022-11-02 fmt --all
> 
> to format the code.
> 
> Also, it looks you need rebase with the master again...

cool just finished the geyser service exit, added "geyser plugin" to cli info, finished merging, and hit it with the format command

lijunwangs commented:
hmm -- this is still out-of-date with the master. Can you do another rebase?

cavemanloverboy commented:
> hmm -- this is still out-of-date with the master. Can you do another rebase?

all done

lijunwangs commented on validator/src/bin/solana-test-validator.rs:
```diff
@@ -559,7 +568,11 @@ fn main() {
+    match genesis.start_with_mint_address(
+        mint_address,
+        socket_addr_space,
+        rpc_to_plugin_manager_receiver,
```
I do not see this parameter being used here. Can you avoid changing this function for general testing as most cases they are not interested in this plugin stuff? If needed, create a new sub variant.

cavemanloverboy replied:
good idea

cavemanloverboy replied:
The ledger tool doesn't need it so I did a similar thing with `GeyserPluginService`

cavemanloverboy commented on ledger-tool/src/main.rs:
```diff
@@ -319,7 +319,7 @@ fn output_ledger(
         .unwrap_or_else(|err| {
             eprintln!("Failed to load entries starting from slot {starting_slot}: {err:?}");
-            exit(1);
+            std::process::exit(1);
```
The variable was named `exit` already. I needed to bring it further up the scope to use it here. When doing so, I needed to either rename the variable or fully qualify the exit function call

lijunwangs replied:
okay then

cavemanloverboy commented on geyser-plugin-manager/src/geyser_plugin_service.rs:
```diff
@@ -14,43 +16,23 @@ use {
     accounts_update_notifier: Option<AccountsUpdateNotifier>,
     transaction_notifier: Option<TransactionNotifierLock>,
     block_metadata_notifier: Option<BlockMetadataNotifierLock>,
+    #[allow(unused)]
```
sounds good. I will name the spawned process as well

lijunwangs commented on test-validator/src/lib.rs:
```diff
@@ -526,7 +531,40 @@ impl TestValidatorGenesis {
         socket_addr_space: SocketAddrSpace,
     ) -> Result<TestValidator, Box<dyn std::error::Error>> {
-        TestValidator::start(mint_address, self, socket_addr_space).map(|test_validator| {
+        TestValidator::start(
```
You can remove this code and forward to start_with_mint_address_and_geyser_plugin_rpc with None for rpc_to_plugin_manager_receiver

cavemanloverboy replied:
done

lijunwangs reviewed (commented):
Looks good to me. Thanks for your contribution and taking time to address the comments. Let's wait for CI's results.

cavemanloverboy commented:
> Looks good to me. Thanks for your contribution and taking time to address the comments. Let's wait for CI's results.

🫡

At some point in the near future I will submit a few more bite sized prs to clean up the unused fields I pointed out earlier (which you wanted as a separate PR) and add a bit more functionality.

lijunwangs commented:
> this is much closer! i left a recommendation to improve encapsulation

@t-nelson Anything else on this PR? Is your concern addressed?

cavemanloverboy commented:
fixed conflicts, will need CI again

lijunwangs commented:
There is a CI failure

lijunwangs commented:
> this is much closer! i left a recommendation to improve encapsulation

@t-nelson @cavemanloverboy  Has this been addressed?

lijunwangs commented:
It is out of sync with master again.

cavemanloverboy commented:
> It is out of sync with master again.

fixed

cavemanloverboy commented:
this is geting out of hand y'all

cavemanloverboy commented:
CI had an issue not on my end

lijunwangs commented:
> CI had an issue not on my end

Retrying

cavemanloverboy commented:
still seems to be an error unrelated to me
```
CXX      google/protobuf/any_lite.lo
--
  | CXX      google/protobuf/arena.lo
  | CXX      google/protobuf/arenastring.lo
  | CXX      google/protobuf/extension_set.lo
  | CXX      google/protobuf/generated_enum_util.lo
  | CXX      google/protobuf/generated_message_table_driven_lite.lo
  | CXX      google/protobuf/generated_message_tctable_lite.lo
  | CXX      google/protobuf/generated_message_util.lo
  | CXX      google/protobuf/implicit_weak_message.lo
  | make[1]: Leaving directory '/var/lib/buildkite-agent/builds/bernal/solana-labs/solana/target/debug/build/protobuf-src-200a799323c91c3b/out/install/build/src'
  |  
  | --- stderr
  | In file included from /usr/include/c++/9/bits/stl_map.h:63,
  | from /usr/include/c++/9/map:61,
  | from /var/lib/buildkite-agent/.cargo/registry/src/github.com-1ecc6299db9ec823/protobuf-src-1.0.5+3.19.3/protobuf/src/google/protobuf/extension_set.h:43,
  | from /var/lib/buildkite-agent/.cargo/registry/src/github.com-1ecc6299db9ec823/protobuf-src-1.0.5+3.19.3/protobuf/src/google/protobuf/extension_set.cc:35:
  | /usr/include/c++/9/tuple:1023:46: internal compiler error: Segmentation fault
  | 1023 \|                     _ConstructibleTuple<_U1, _U2>()
  | \|                                              ^~~
  | Please submit a full bug report,
  | with preprocessed source if appropriate.
  | See <file:///usr/share/doc/gcc-9/README.Bugs> for instructions.
  | make[1]: *** [Makefile:4503: google/protobuf/extension_set.lo] Error 1
  | make[1]: *** Waiting for unfinished jobs....
  | In file included from /var/lib/buildkite-agent/.cargo/registry/src/github.com-1ecc6299db9ec823/protobuf-src-1.0.5+3.19.3/protobuf/src/google/protobuf/generated_message_tctable_lite.cc:36:
  | /var/lib/buildkite-agent/.cargo/registry/src/github.com-1ecc6299db9ec823/protobuf-src-1.0.5+3.19.3/protobuf/src/google/protobuf/generated_message_tctable_impl.h: In function ‘void google::protobuf::internal::AlignFail(uintptr_t) [with long unsigned int align = 4]’:
  | /var/lib/buildkite-agent/.cargo/registry/src/github.com-1ecc6299db9ec823/protobuf-src-1.0.5+3.19.3/protobuf/src/google/protobuf/generated_message_tctable_impl.h:103:1: warning: ‘noreturn’ function does return
  | 103 \| }
  | \| ^
  | /var/lib/buildkite-agent/.cargo/registry/src/github.com-1ecc6299db9ec823/protobuf-src-1.0.5+3.19.3/protobuf/src/google/protobuf/generated_message_tctable_impl.h: In function ‘void google::protobuf::internal::AlignFail(uintptr_t) [with long unsigned int align = 8]’:
  | /var/lib/buildkite-agent/.cargo/registry/src/github.com-1ecc6299db9ec823/protobuf-src-1.0.5+3.19.3/protobuf/src/google/protobuf/generated_message_tctable_impl.h:103:1: warning: ‘noreturn’ function does return
  | make: *** [Makefile:1913: install-recursive] Error 1
  | thread 'main' panicked at '
  | command did not execute successfully, got: exit status: 2
  |  
  | build script failed, must exit now', /var/lib/buildkite-agent/.cargo/registry/src/github.com-1ecc6299db9ec823/autotools-0.2.5/src/lib.rs:715:5
  | stack backtrace:
  | 0: rust_begin_unwind
  | at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/panicking.rs:575:5
  | 1: core::panicking::panic_fmt
  | at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/panicking.rs:64:14
  | 2: autotools::fail
  | at /var/lib/buildkite-agent/.cargo/registry/src/github.com-1ecc6299db9ec823/autotools-0.2.5/src/lib.rs:715:5
  | 3: autotools::run
  | at /var/lib/buildkite-agent/.cargo/registry/src/github.com-1ecc6299db9ec823/autotools-0.2.5/src/lib.rs:686:9
  | 4: autotools::Config::build
  | at /var/lib/buildkite-agent/.cargo/registry/src/github.com-1ecc6299db9ec823/autotools-0.2.5/src/lib.rs:662:9
  | 5: build_script_build::main
  | at ./build.rs:26:5
  | 6: core::ops::function::FnOnce::call_once
  | at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/ops/function.rs:507:5
  | note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
  | warning: build failed, waiting for other jobs to finish...
  | 🚨 Error: The command exited with status 101
```

lijunwangs commented:
ahh. It is out-of-date with master again? can you do another rebase?

CriesofCarrots commented:
@lijunwangs , the "out-of-date" message isn't a big deal, if there aren't merge conflicts.

lijunwangs reviewed (approved):
LGTM

behzadnouri commented:
this has broken master branch

cavemanloverboy commented:
> this has broken master branch

what are you running into?

behzadnouri commented:
> > this has broken master branch
> 
> what are you running into?

https://buildkite.com/solana-labs/solana/builds/91996#0186ece5-27ed-4203-b932-e66978358b02

cavemanloverboy commented:
> > > this has broken master branch
> > 
> > 
> > what are you running into?
> 
> https://buildkite.com/solana-labs/solana/builds/91996#0186ece5-27ed-4203-b932-e66978358b02

this PR didn't touch that line, pretty sure

CriesofCarrots commented:
> this PR didn't touch that line

No, those lines were added in #30755 , and did need to be updated in your PR in order to compile.
We don't see a need for the ledger-tool changes, though, so are reverting them: https://github.com/solana-labs/solana/pull/30763
Can you please let me know if we missed something?

lijunwangs commented:
> > this PR didn't touch that line
> 
> No, those lines were added in #30755 , and did need to be updated in your PR in order to compile. We don't see a need for the ledger-tool changes, though, so are reverting them: #30763 Can you please let me know if we missed something?

Wondering why we did not catch it in CI before the merge...

apfitzge commented on geyser-plugin-manager/src/geyser_plugin_manager.rs:
```diff
@@ -20,26 +21,6 @@ impl GeyserPluginManager {
-        config_file: &str,
-    ) -> Result<(), Box<dyn Error>> {
-        type PluginConstructor = unsafe fn() -> *mut dyn GeyserPlugin;
-        let lib = Library::new(libpath)?;
-        let constructor: Symbol<PluginConstructor> = lib.get(b"_create_plugin")?;
-        let plugin_raw = constructor();
-        let mut plugin = Box::from_raw(plugin_raw);
```
cc/ @OliverNChalk
bug introduced by the code that replaced this.

In Rust variables on stack will be dropped in reverse declaration order.
In this (removed) version of the function, library lives longer than the plugin.
In the new version, library is dropped first which will lead to sigsev.

cavemanloverboy replied:
rip
